org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentHighlightHandler.java

```diff
@@ -18,75 +18,138 @@
+ */
+public class DocumentHighlightHandler {
+
+	/**
+	 * Handles a {@code textDocument/documentHighlight} request.
+	 *
+	 * @param params the position at which to find highlights
+	 * @param monitor the progress monitor
+	 * @return the document highlights for the given position
+	 */
+	public static List<DocumentHighlight> documentHighlight(TextDocumentPositionParams params, IProgressMonitor monitor) {
+		ITypeRoot typeRoot = JDTUtils.resolveTypeRoot(params.getTextDocument().getUri());
+		if (typeRoot == null || monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
+		CompilationUnit ast = CoreASTProvider.getInstance().getAST(typeRoot, CoreASTProvider.WAIT_YES, monitor);
+		if (ast == null || monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
+
+		int offset = JsonRpcHelpers.toOffset(typeRoot,
+			params.getPosition().getLine(), params.getPosition().getCharacter());
+		ASTNode node = NodeFinder.perform(ast, offset, 0);
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
+
+		return findHighlights(ast, node, monitor);
+	}
+
+	/**
+	 * Finds {@link DocumentHighlight}s in a {@link CompilationUnit}.
+	 * The highlights are searched using the following {@link IOccurrencesFinder}s:
+	 * <ol>
+	 *   <li>{@link ExceptionOccurrencesFinder}</li>
+	 *   <li>{@link MethodExitsFinder}</li>
+	 *   <li>{@link BreakContinueTargetFinder}</li>
+	 *   <li>{@link ImplementOccurrencesFinder}</li>
+	 *   <li>{@link OccurrencesFinder}</li>
+	 * </ol>
+	 *
+	 * @param ast the {@link CompilationUnit}
+	 * @param node the selected {@link ASTNode} to find highlights for
+	 * @param monitor the progress monitor
+	 * @return the highlights, or an empty list if none were found
+	 */
+	private static List<DocumentHighlight> findHighlights(CompilationUnit ast, ASTNode node, IProgressMonitor monitor) {
+		IOccurrencesFinder finder;
+
+		finder = new ExceptionOccurrencesFinder();
+		if (finder.initialize(ast, node) == null) {
+			return convertToHighlights(ast, finder.getOccurrences());
+		}
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
+
+		finder = new MethodExitsFinder();
+		if (finder.initialize(ast, node) == null) {
+			return convertToHighlights(ast, finder.getOccurrences());
 		}
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
+
+		finder = new BreakContinueTargetFinder();
+		if (finder.initialize(ast, node) == null) {
+			return convertToHighlights(ast, finder.getOccurrences());
+		}
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
+
+		finder = new ImplementOccurrencesFinder();
+		if (finder.initialize(ast, node) == null) {
+			return convertToHighlights(ast, finder.getOccurrences());
+		}
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
+
+		finder = new OccurrencesFinder();
+		if (finder.initialize(ast, node) == null) {
+			return convertToHighlights(ast, finder.getOccurrences());
+		}

```

Comment:
I was thinking the same, the only downside with doing what you suggest is the need to instantiate all finders. My original code does it the same [way as it's done in JDT.UI](https://git.eclipse.org/r/plugins/gitiles/jdt/eclipse.jdt.ui/+/1b6550aba78bcd9ad37699f19ce4bef6b812c8b0/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/JavaEditor.java#3314), which (sometimes) avoids a few instantiations when occurrences are found by one of the earlier finders. It shouldn't really make a difference though since all finders are usually needed anyways, and all basically all initialization is done in `IOccurencesFinder#initialize`.

I can refactor this when I take the time to work on this PR again.

Reply:
Ok, in that case I'm fine keeping it as is. I suppose I got a aggressive on keeping things small when I saw the size of the PR but looking back we may be able to reduce a lot of that.